**1. Hava durumu modeli**

Bu ödev için, geçmiş sıcaklık veri kümesine yeniden bakacaksınız. Aşağıdaki görevler için çözümünüzü içeren Jupyter notebooks'u gist dosyasına link olarak ekleyin

In [ ]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

* Önceki derste olduğu gibi, hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.

In [ ]:
import statsmodels.api as sm
weather_df["Yeni"] = weather_df["Apparent Temperature (C)"] - weather_df["Temperature (C)"]
Y = weather_df["Yeni"]
X = weather_df[["Humidity", "Wind Speed (km/h)"]]
X = sm.add_constant(X)
sonuclar1 = sm.OLS(Y, X).fit()
sonuclar1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   Yeni   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        15:24:09   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-kare değeri 0.288 ve Adj. R-kare değeri 0.288 dir. Çok düşük bir sonuçtur. Sonucun 1 e yakın olması gerekiyor.

* Daha sonra, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Şimdi, bu modelin R-kare değeri nedir? Bu model bir öncekine göre gelişti mi?

In [ ]:
Y = weather_df["Yeni"]
weather_df['hum_wind'] = weather_df.Humidity * weather_df["Wind Speed (km/h)"]
X = weather_df[["Humidity", "Wind Speed (km/h)","hum_wind"]]

X = sm.add_constant(X)

sonuclar2 = sm.OLS(Y, X).fit()
sonuclar2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   Yeni   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        15:24:10   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.0839      0.033      2.511      0.012       0.018       0.149
Humidity              0.1775      0.043      4.133      0.000       0.093       0.262
Wind Speed (km/h)     0.0905      0.002     36.797      0.000       0.086       0.095
hum_wind             -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-Kare ve Adj. R-Kare değerleri 0.341 dir ve Önceki sonuca göre artmıştır.

* İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?

In [ ]:
weather_df["Yeni"] = weather_df["Apparent Temperature (C)"] - weather_df["Temperature (C)"]
Y = weather_df["Yeni"]
X = weather_df[["Humidity", "Wind Speed (km/h)", "Visibility (km)"]]
X = sm.add_constant(X)
sonuclar3 = sm.OLS(Y, X).fit()
sonuclar3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   Yeni   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        15:24:12   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 1.5756      0.028     56.605      0.000       1.521       1.630
Humidity             -2.6066      0.025   -102.784      0.000      -2.656      -2.557
Wind Speed (km/h)    -0.1199      0.001   -179.014      0.000      -0.121      -0.119
Visibility (km)       0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-Kare 0.304 bulundu. Adj.R-Kare değeri 0.303 dür. İlk Modele göre sonuçlar daha iyidir. Fakat etkileşimli modele göre daha düşük R-kare değerlerine sahiptir. Etkileşimli model daha kullanışlıdır.

* AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.

1.Model AIC = 3.409e+05 BIC = 3.409e+05

2.Model AIC = 3.334e+05 BIC = 3.334e+05

3.Model AIC = 3.388e+05 BIC =	3.388e+05

AIC ve BIC değerlerinin düşük olması modelin daha iyi sonuç verdiğini gösterdiğinden 2.model seçilmeli

2. Ev fiyatları modeli
Bu alıştırmada, ev fiyatları modelinizde çalışacaksınız. Aşağıdaki görevler için çözümünüzü içeren Jupyter notebooks'u gist dosyasına link olarak ekleyin.:

In [ ]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

* Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.

In [ ]:
import statsmodels.api as sm
house_prices_df = house_prices_df.dropna(axis=1, how="any")
Y = house_prices_df["SalePrice"]
X = house_prices_df[["OverallQual", "TotalBsmtSF", "GrLivArea", "GarageCars", "GarageArea"]]
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     926.5
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        16:34:19   Log-Likelihood:                -17499.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1454   BIC:                         3.504e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.907e+04   4638.450    -21.359      0.000   -1.08e+05      -9e+04
OverallQual  2.364e+04   1072.532     22.037      0.000    2.15e+04    2.57e+04
TotalBsmtSF    31.5015      2.904     10.848      0.000      25.805      37.198
GrLivArea      45.3458      2.489     18.218      0.000      40.463      50.228
GarageCars   1.454e+04   3022.681      4.812      0.000    8615.034    2.05e+04
GarageArea     17.1334     10.468      1.637      0.102      -3.401      37.668
==============================================================================
Omnibus:                      438.146   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            40648.199
Skew:                          -0.338   Prob(JB):                         0.00
Kurtosis:                      28.841   Cond. No.                     9.23e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.23e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared:	0.761 - Adj. R-squared:	0.760 - F-statistic:	926.5 - 
AIC:	3.501e+04 - BIC:	3.504e+04

* Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın

Bulunan değerlendirme metriklerine göre sonuçlar ortalama olarak bulunmuştur. R değerleri ve F değeri daha yüksek olabilirdi. AIC ve BIC değerleri de daha düşük bulunabilirdi. Modelin biraz daha geliştirilmesi lazım.

* Modelinizin uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.

In [ ]:
house_prices_df = house_prices_df.dropna(axis=1, how="any")
Y = house_prices_df["SalePrice"]
house_prices_df["New"] = house_prices_df.OverallQual * house_prices_df.GrLivArea
X = house_prices_df[["OverallQual", "TotalBsmtSF", "GrLivArea", "GarageCars", "GarageArea","New"]]
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

Etkileşimli değişken eklendi. Modelin metrik sonuçları biraz daha düzeldi.

* Denediğiniz her model için, uygun metrikleri alın ve modellerinizi birbiriyle karşılaştırın. Hangi model en iyisidir ve neden?

In [ ]:
house_prices_df = house_prices_df.dropna(axis=1, how="any")
Y = house_prices_df["SalePrice"]
X = house_prices_df[["OverallQual", "TotalBsmtSF", "GrLivArea", "GarageCars"]]
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     1156.
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        16:42:20   Log-Likelihood:                -17500.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1455   BIC:                         3.504e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.925e+04   4639.866    -21.390      0.000   -1.08e+05   -9.01e+04
OverallQual  2.357e+04   1072.465     21.980      0.000    2.15e+04    2.57e+04
TotalBsmtSF    32.5203      2.838     11.459      0.000      26.953      38.087
GrLivArea      45.6432      2.484     18.376      0.000      40.771      50.515
GarageCars   1.858e+04   1747.412     10.634      0.000    1.52e+04     2.2e+04
==============================================================================
Omnibus:                      421.692   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            36798.858
Skew:                          -0.273   Prob(JB):                         0.00
Kurtosis:                      27.589   Cond. No.                     8.95e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.95e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
house_prices_df = house_prices_df.dropna(axis=1, how="any")
Y = house_prices_df["SalePrice"]
house_prices_df["New"] = house_prices_df.OverallQual * house_prices_df.GrLivArea
X = house_prices_df[["OverallQual", "TotalBsmtSF", "GrLivArea", "GarageCars", "GarageArea","New"]]
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.776
Model:                            OLS   Adj. R-squared:                  0.775
Method:                 Least Squares   F-statistic:                     839.0
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        16:42:33   Log-Likelihood:                -17452.
No. Observations:                1460   AIC:                         3.492e+04
Df Residuals:                    1453   BIC:                         3.495e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        3755.3845   1.14e+04      0.330      0.742   -1.86e+04    2.61e+04
OverallQual  8189.5922   1883.684      4.348      0.000    4494.561    1.19e+04
TotalBsmtSF    27.2851      2.845      9.589      0.000      21.704      32.867
GrLivArea     -25.2067      7.572     -3.329      0.001     -40.059     -10.354
GarageCars   1.905e+04   2963.564      6.427      0.000    1.32e+04    2.49e+04
GarageArea      3.7542     10.231      0.367      0.714     -16.315      23.824
New            10.4535      1.063      9.830      0.000       8.367      12.540
==============================================================================
Omnibus:                     1102.923   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           173172.934
Skew:                          -2.628   Prob(JB):                         0.00
Kurtosis:                      56.095   Cond. No.                     1.30e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.3e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
house_prices_df = house_prices_df.dropna(axis=1, how="any")
Y = house_prices_df["SalePrice"]
house_prices_df["New"] = house_prices_df.TotalBsmtSF * house_prices_df.GrLivArea
X = house_prices_df[["OverallQual", "TotalBsmtSF", "GrLivArea", "GarageCars", "GarageArea","New"]]
X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.769
Model:                            OLS   Adj. R-squared:                  0.768
Method:                 Least Squares   F-statistic:                     807.9
Date:                Sun, 05 Dec 2021   Prob (F-statistic):               0.00
Time:                        16:42:34   Log-Likelihood:                -17473.
No. Observations:                1460   AIC:                         3.496e+04
Df Residuals:                    1453   BIC:                         3.500e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.256e+05   5856.973    -21.448      0.000   -1.37e+05   -1.14e+05
OverallQual  2.272e+04   1061.805     21.394      0.000    2.06e+04    2.48e+04
TotalBsmtSF    59.7732      4.846     12.336      0.000      50.268      69.278
GrLivArea      63.8004      3.538     18.033      0.000      56.860      70.741
GarageCars   1.034e+04   3027.446      3.415      0.001    4400.681    1.63e+04
GarageArea     26.4354     10.369      2.549      0.011       6.095      46.776
New            -0.0133      0.002     -7.220      0.000      -0.017      -0.010
==============================================================================
Omnibus:                      534.220   Durbin-Watson:                   1.962
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            11176.535
Skew:                           1.184   Prob(JB):                         0.00
Kurtosis:                      16.346   Cond. No.                     1.32e+07
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.32e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Etkileşimli değer kullanılan 2.model daha iyi sonuçlar verdi. Bunun içinde AIC BIC, f score ve R değerlerine bakıldı. Buna göre karar verildi.